### PR TITLE
Docker Desktop support

### LINKED ISSUE #321: WSL issues
WSL might not be completely compatible with the plugin.

Can host.docker.internal be used on Linux? Handling localhost uniformly would simplify the code and potentially prevent issues.
Possible solution, didn't get it to work locally on my Linux machine: https://stackoverflow.com/questions/30109037/how-can-i-forward-localhost-port-on-my-container-to-localhost-on-my-host

### PR DESCRIPTION
**Pull request by notV4l**
## Usage related changes

- Close #321 

## Checklist:

-   [x] Formatted the code
-   [x] No linter errors + tried to avoid introducing linter warnings
-   [x] Performed a self-review of the code
-   [ ] Rebased to the last commit of the target branch (or merged it into my branch)
-   [ ] Documented the changes
-   [ ] Updated the `test` directory (with a test case consisting of `network.json`, `hardhat.config.ts`, `check.ts`)
-   [x] Linked issues which this PR resolves
-   [ ] Created a PR to the `plugin` branch of [`starknet-hardhat-example`](https://github.com/Shard-Labs/starknet-hardhat-example):
    -   < EXAMPLE_REPO_PR_URL > 
    -   [ ] Modified `test.sh` to use my example repo branch
    -   [ ] Restored `test.sh` to to use the original branch (after the example repo PR has been merged)
-   [ ] All tests are passing (for external contributors who don't have access to the CI/CD pipeline)

### CODE REVIEW

File: src/external-server/docker-server.ts [hunk 1]
```diff
@@ -46,9 +47,15 @@ export abstract class DockerServer extends ExternalServer {
             formattedImage,
             ...(await this.getContainerArgs())
         ];
+
         return spawn("docker", args);
     }
 
+    public isDockerDesktop(): boolean {
```
Comment (FabijanC) [inline src/external-server/docker-server.ts:54]: I would implement this as a cached [JS getter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get) (does it have to be public?) which first checks a private property to see if it was already executed.

File: src/external-server/external-server.ts [hunk 2]
```diff
@@ -149,6 +149,7 @@ export abstract class ExternalServer {
     }
 
     private async isServerAlive() {
+        if (this.port === null) return false;
```
Comment (FabijanC) [inline src/external-server/external-server.ts:152]: Why is this necessary?

File: src/utils.ts [hunk 3]
```diff
@@ -53,14 +53,13 @@ export function adaptLog(msg: string): string {
...
-const MACOS_PLATFORM = "darwin";
 /**
- * Adapts `url` by replacing localhost and 127.0.0.1 with `host.internal.docker`
+ * Adapts `url` by replacing localhost and 127.0.0.1 with `host.docker.internal`
  * @param url string representing the url to be adapted
  * @returns adapted url
  */
-export function adaptUrl(url: string): string {
-    if (process.platform === MACOS_PLATFORM) {
+export function adaptUrl(url: string, isDockerDesktop: boolean): string {
```
Comment (FabijanC) [inline src/utils.ts:61]: Instead of having to provide the boolean argument, could this `adaptUrl` function find out by itself if the Docker is indeed a desktop one?

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: src/external-server/docker-server.ts (hunk 1 above)
Comment (notV4l) [reply src/external-server/docker-server.ts:54]: its only executed once when spawning docker container

File: src/external-server/external-server.ts (hunk 2 above)
Comment (notV4l) [reply src/external-server/external-server.ts:152]: starknet-docker-proxy is initialized with null port, it avoid one request with null port

**Review by Nathan-SL — APPROVED** [review]

File: src/external-server/external-server.ts [hunk 4]
```diff
@@ -57,6 +59,11 @@ export abstract class ExternalServer {
         protected stderr?: string
     ) {
         ExternalServer.cleanupFns.push(this.cleanup.bind(this));
+        this._isDockerDesktop = isDockerDesktop();
+    }
+
+    public get isDockerDesktop(): boolean {
```
Comment (FabijanC) [inline src/external-server/external-server.ts:65]: I didn't mean the getter to just return the value of a private property, but also set the value of it so that it doesn't have to be set elsewhere. And so that this getter is the only place responsible for checking.

File: src/utils.ts [hunk 5]
```diff
@@ -77,6 +77,13 @@ export function adaptUrl(url: string): string {
 
     return url;
 }
+/**
+ * Check if docker is Docker Desktop
+ */
+export function isDockerDesktop(): boolean {
+    const res = spawnSync("docker", ["system", "info"], { encoding: "utf8" });
+    return res?.stdout?.includes("Operating System: Docker Desktop");
```
Comment (FabijanC) [inline src/utils.ts:85]: Why the question marks? Shouldn't these properties always exist? When I earlier said that this looks flaky, I meant that we are depending on the output format of the `docker system info command`, which might change or not be the same on all platforms.

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: src/utils.ts (hunk 3 above)
Comment (notV4l) [reply src/utils.ts:62]: i guess this logic was in utils.ts to make it accessible from anywhere, that's also why i moved isDockerDesktop function to utils.ts

File: src/starknet-docker-proxy.ts [hunk 6]
```diff
@@ -28,8 +28,10 @@ export class StarknetDockerProxy extends DockerServer {
     }
 
     protected async getDockerArgs(): Promise<string[]> {
-        // To access the files on host machine from inside the container, proper mounting has to be done.
+        // get a valid free port
+        this.port = await this.getPort();
```
[the remark this answers is not in the document]
Comment (notV4l) [reply src/starknet-docker-proxy.ts:32]: cause StarknetDockerProxy is initialized with port = null

File: src/starknet-wrappers.ts [hunk 7]
```diff
@@ -446,14 +449,16 @@ export class DockerWrapper extends StarknetWrapper {
         cairoPaths: string[],
         hre: HardhatRuntimeEnvironment
     ) {
-        super(new StarknetDockerProxy(image, rootPath, accountPaths, cairoPaths), hre);
+        const externalServer = new StarknetDockerProxy(image, rootPath, accountPaths, cairoPaths);
+        super(externalServer, hre);
+        this.externalServer = externalServer;
```
[the remark this answers is not in the document]
Comment (notV4l) [reply src/starknet-wrappers.ts:454]: indeed

File: src/starknet-docker-proxy.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (FabijanC) [reply src/starknet-docker-proxy.ts:32]: And it's initialized with null to save one request?

File: src/utils.ts (hunk 3 above)
Comment (FabijanC) [reply src/utils.ts:62]: `adaptUrl` was in utils because it really was used in multiple places. Now it's used in only one place (after a refactor was done a few months ago).

File: src/utils.ts (hunk 5 above)
Comment (notV4l) [reply src/utils.ts:85]: this morning i had to add question marks to make tests pass while it was ok yesterday :|

File: src/utils.ts (hunk 3 above)
Comment (notV4l) [reply src/utils.ts]: refactored

File: src/starknet-wrappers.ts [hunk 8]
```diff
@@ -486,10 +502,6 @@ export class VenvWrapper extends StarknetWrapper {
         super(new StarknetVenvProxy(pythonPath), hre);
     }
 
-    protected override get gatewayUrl(): string {
```
Comment (FabijanC) [inline src/starknet-wrappers.ts:489]: This shouldn't be removed. The adaptation only needs to take place if using DockerWrapper. Why would we adapt the URL if the user has Docker Desktop if they aren't even using the plugin in dockerized mode?

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: src/starknet-wrappers.ts (hunk 8 above)
Comment (notV4l) [reply src/starknet-wrappers.ts:489]: my bad

File: src/utils.ts (hunk 5 above)
Comment (notV4l) [reply src/utils.ts]: yeah :) but did u know https://www.bbc.com/future/article/20221011-how-space-weather-causes-computer-errors XD

Comment (FabijanC) [reply src/utils.ts]: that's it, I no longer need excuses when I can blame the universe xD

**Review by FabijanC — COMMENTED** [review]
Looks better now, but we still have one unresolved conversation.

File: src/starknet-docker-proxy.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (notV4l) [reply src/starknet-docker-proxy.ts]: to get a free port dynamically, i moved this call just before using this.port

**Review by FabijanC — APPROVED** [review]
